controllers/topology/machinedeployment_controller.go

```diff
@@ -143,10 +144,35 @@ func (r *MachineDeploymentReconciler) reconcileDelete(ctx context.Context, md *c
 	if err != nil {
 		return ctrl.Result{}, errors.Wrapf(err, "failed to create patch helper for %s", tlog.KObj{Obj: md})
 	}
+
+	// If the MachineDeployment has a MachineHealthCheck delete it.
+	if err := r.deleteMachineHealthCheckForMachineDeployment(ctx, md); err != nil {
+		return ctrl.Result{}, err
+	}
+
 	controllerutil.RemoveFinalizer(md, clusterv1.MachineDeploymentTopologyFinalizer)
 	if err := patchHelper.Patch(ctx, md); err != nil {
 		return ctrl.Result{}, errors.Wrapf(err, "failed to patch %s", tlog.KObj{Obj: md})
 	}
 
 	return ctrl.Result{}, nil
 }
+
+func (r *MachineDeploymentReconciler) deleteMachineHealthCheckForMachineDeployment(ctx context.Context, md *clusterv1.MachineDeployment) error {
+	mhc := &clusterv1.MachineHealthCheck{}
+
+	// MachineHealthCheck always has the same name and namespace as the MachineDeployment it belongs to.
+	err := r.Client.Get(ctx, client.ObjectKey{Namespace: md.Namespace, Name: md.Name}, mhc)
+	if err != nil {
+		// Each MachineDeployment isn't required to have a MachineHealthCheck. Ignore the error if it's of the type not found, but return any other error.
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return errors.Wrap(err, fmt.Sprintf("%s MachineHealthCheckObject could not be retrieved for", tlog.KObj{Obj: md}))
+	}
+	if err := r.Client.Delete(ctx, mhc); err != nil {
```

Comment:
Can we maybe just delete the MHC by constructing a MHC with the right name/namespace, delete it and then ignore the error if it doesn't exist? (and thus avoid the get entirely)

Reply:
Good idea!